core/src/test/java/io/questdb/test/cutlass/http/line/LineHttpSenderFailureTest.java

```diff
@@ -0,0 +1,139 @@
+
+public class LineHttpSenderFailureTest extends AbstractBootstrapTest {
+
+    public static void scenarioRetryWithDeduplication(ServerController controller) throws Exception {
+        String tableName = UUID.randomUUID().toString();
+        TestUtils.assertMemoryLeak(() -> {
+            try {
+                controller.startAndExecute("create table '" + tableName + "' (value long, ts timestamp) timestamp (ts) partition by DAY WAL DEDUP UPSERT KEYS(ts)");
+                CountDownLatch senderLatch = new CountDownLatch(2); // one for Sender and one for Restarter
+
+                AtomicReference<Exception> senderException = new AtomicReference<>();
+                new Thread(() -> {
+                    try (Sender sender = controller.newSender()) {
+                        for (int i = 0; i < 1_000_000; i++) {
+                            sender.table(tableName).longColumn("value", 42).at(i * 10, ChronoUnit.MICROS);
+                        }
+                    } catch (Exception t) {
+                        senderException.set(t);
+                    } finally {
+                        senderLatch.countDown();
+                    }
+                }).start();
+
+                new Thread(() -> {
+                    // keeping restarting the server until Sender is done
+                    while (senderLatch.getCount() == 2) {
+                        Os.sleep(500);
+                        controller.restart();
+                    }
+                    controller.stop(); // stop will clear the thread local Path
+                    senderLatch.countDown();
+                }).start();
+
+                senderLatch.await();
+
+                if (senderException.get() != null) {
+                    Assert.fail("Sender failed: " + senderException.get().getMessage());
+                }
+                controller.start();
+                controller.assertSqlEventually("select count() from '" + tableName + "'", "count\n1000000\n");
+            } finally {
+                Misc.free(controller);
+            }
+        });
+    }
+
+    @Test
+    public void scenarioRetryWithDeduplication() throws Exception {
+        scenarioRetryWithDeduplication(new ServerController());
+    }
+
+    @Before
+    public void setUp() {
+        super.setUp();
+        TestUtils.unchecked(() -> createDummyConfiguration());
+        dbPath.parent().$();
+    }
+
+    public static class ServerController implements Closeable {
+        private TestServerMain serverMain;
+
+        public void assertSqlEventually(String sql, String expected) {
+            TestUtils.assertEventually(() -> serverMain.assertSql(sql, expected));
+        }
+
+        @Override
+        public void close() {
+            serverMain = Misc.free(serverMain);
+        }
+
+        public Sender newSender() {
+            String url = "http://localhost:" + HTTP_PORT;
+            return Sender.builder().url(url).maxPendingRows(100).retryTimeoutMillis(1000).build();
```

Comment:
nit: I'd use larger retry timeout since some of the CI workers are quite slow.

Reply:
good one. I will also increase the batch size. to make the test a bit faster.